00-programming-fundamentals/intro-to-regex.md

```diff
@@ -0,0 +1,133 @@
+
+A range of repetitions can also be repeated by using two parameters in the curly braces.  For example `/[abc]{3, 5}/` would allow the characters to repeat between 3 and 5 times.
+
+## Capture Groups
+
+We can also construct groups of characters which can be combined with special characters using the parentheses.  For example `/(ada *)+/` would match one or more "ada" lines separated by 0 or more spaces.  So "ada ada", "ada", and "ada ada   ada" would all match.
+
+So `/a(ab)*a/` would match the strings:  "aaba", "aa", "aababa" because it requires an a followed by zero or more "ab" blocks and ending with an "a".
+
+Similarly, `/\$(\d)+\.(\d){2}` would match us currency, like you were asked to deduce above requiring a dollar sign followed by one or more digits, followed by a period and ending with 2 digits.  
+
+
+
+### Practice
+
+See if you can write a regular expression for these patterns:
+
+1. A 5 digit number
+1. An IP address like 127.0.0.1, or 206.190.36.45
+
+[Sample Solutions](https://gist.github.com/CheezItMan/54e39aa4a355aa5a24bb281d5d360121)
+
+## List of Regex Special Characters
+
+| Character | Name  |  Description |
+| --- | --- | --- |
+|   \	|   Backslash	|   The backslash gives special meaning to the character following it. For example, the combination "\n" stands for the newline, one of the control characters. 	|
+|   ^	|   Caret	|   The caret is the start of line anchor or the negate symbol.  Example: "^a" matches "a" at the start of a line.  Example: "[^0-9]" matches any non digit.	|
+|   $	|   Dollar	|   $ the dollar is the end of line anchor.	|
+|   {}	|   Curly Braces	|   { } the open and close curly bracket are used as range quantifiers.	|
+|   []	|   Square Brackets	|    Open and close square bracket define a character class to match a single character.	|
+|   ()	|   Parentheses 	|   The open and close parenthesis are used for grouping characters	|
+|   .	|   Dot	|   the dot matches any character except the newline.	|
+|   *	|   Star	|   The star is the match-zero-or-more quantifier.	|
+|   +	|   Plus	|   The plus is the match-zero-or-one quantifier.	|
+|   ?	|   Question Mark	|   The question mark	 is the match-one-or-more quantifier.	|
+|   &#124;	|   Pipe	|   The vertical pipe separates a series of alternatives.	|
+|  < > 	|   Anchors	|   The smaller and greater signs are anchors that specify a left or right word boundary.	|
+|  - 	|   Minus	|   the minus indicates a range in a character class (when it is not at the first position after the "[" opening bracket or the last position before the "]" closing bracket. For example "[A-Z]" matches any uppercase character.	|
+| & | Ampersand | The and is the "substitute complete match" symbol. |
+
+
```

Comment:
I think it would be good to have a conclusion that highlights:
- Regex is available in all languages, but sometimes bits of the syntax are different
- There are many tools to help write and validate regex expressions, like http://rubular.com/
- You can look up many common regex expressions, but it is important to know the basics to you can break down each piece

Reply:
Added